### PR TITLE
New spike report API

### PR DESCRIPTION
**Pull request by kaabimg**
Please review this pull request.
Do not merge it since that the brain API is broken.

### CODE REVIEW

**Comment by hernando** [comment]
I didn't finish with the review, but until Monday I can't continue.

File: apps/monsteer_spike_receiver.cpp [hunk 1]
```diff
@@ -54,20 +54,20 @@ class SpikeReceiver
...
-        while( _reader.waitUntil( time ))
-        {
-            const brion::Spikes& spikes = _reader.getSpikes();
-            BOOST_FOREACH( const brion::Spike& spike, spikes )
-                std::cout << spike.first << " " << spike.second << std::endl;
+//        float time = 0;
+//        while( _reader.waitUntil( time ))
+//        {
+//            const brion::Spikes& spikes = _reader.getSpikes();
+//            BOOST_FOREACH( const brion::Spike& spike, spikes )
```
Comment (kaabimg) [inline apps/monsteer_spike_receiver.cpp]: This code was commented. The implementation is the same as Brion/apps/spikeConverter
May be that it should be removed.

File: apps/monsteer_spike_receiver.cpp [hunk 2]
```diff
@@ -54,20 +54,20 @@ class SpikeReceiver
...
-        /* Printing the last window received */
-        const brion::Spikes& spikes = _reader.getSpikes();
-        BOOST_FOREACH( const brion::Spike& spike, spikes )
-            std::cout << spike.first << " " << spike.second << std::endl;
+//            _reader.clear( 0, time );
+//            time = time + MUSIC_TIMESTEP;
+//        }
+//        /* Printing the last window received */
+//        const brion::Spikes& spikes = _reader.getSpikes();
+//        BOOST_FOREACH( const brion::Spike& spike, spikes )
```
Comment (kaabimg) [inline apps/monsteer_spike_receiver.cpp]: Same comment as above.

File: apps/monsteer_spike_receiver.cpp [hunk 3]
```diff
@@ -54,20 +54,20 @@ class SpikeReceiver
...
+//            std::cout << spike.first << " " << spike.second << std::endl;
```
Comment (kaabimg) [inline apps/monsteer_spike_receiver.cpp]: Same comment as above.

File: monsteer/plugin/spikeReport.cpp [hunk 4]
```diff
@@ -20,76 +19,78 @@
...
+{
+    return BRION_VERSION_ABI;
+}
 extern "C" bool LunchboxPluginRegister()
 {
     lunchbox::PluginRegisterer< monsteer::plugin::SpikeReport > registerer;
     return true;
 }
 
+#define RECEIVE_TIMEOUT 100 // ms
```
Comment (kaabimg) [inline monsteer/plugin/spikeReport.cpp]: This is a legacy code that I didn't really touched. I don't know why it is there in this form.

File: monsteer/plugin/spikeReport.cpp [hunk 5]
```diff
@@ -98,178 +99,174 @@ std::string SpikeReport::getDescription()
...
+        _currentTime = brion::UNDEFINED_TIMESTAMP;
+        _state = State::ended;
+        _spikes.clear();
+        return;
     }
-    return _lastTimeStamp;
+
+    auto position = std::upper_bound(
+        _spikes.begin(), _spikes.end(), toTimeStamp,
+        []( float val, const brion::Spike &spike ) { return spike.first >= val; } );
```
Comment (hernando) [inline monsteer/plugin/spikeReport.cpp]: To me this is clearer is you return val < spike.first

File: monsteer/plugin/spikeReport.h [hunk 6]
```diff
@@ -21,73 +21,53 @@
...
+    virtual const URI &getURI() const;
 
     void close() final;
+    brion::Spikes read( float min ) final;
+    brion::Spikes readUntil( float max ) final;
+    void readSeek( float toTimeStamp ) final;
+    void writeSeek( float toTimeStamp ) final;
+    void write( const brion::Spikes &spikes ) final;
 
 private:
```
Comment (kaabimg) [inline monsteer/plugin/spikeReport.h]: Done

**Pull request merged**